packages/types/src/transaction/get-stub-action-view-from-plan.ts

```diff
@@ -0,0 +1,144 @@
+  if (!value.assetId) throw new Error('No asset ID in value');
+  if (!value.amount) throw new Error('No amount in value');
+
+  const denomMetadata = denomMetadataByAssetId[uint8ArrayToBase64(value.assetId.inner)];
+  if (!denomMetadata) throw new Error('Asset ID in spend plan refers to an unknown asset type');
+
+  return new ValueView({
+    valueView: {
+      case: 'knownDenom',
+      value: {
+        amount: value.amount,
+        denom: denomMetadata,
+      },
+    },
+  });
+};
+
+const getNoteView = (note: Note | undefined, denomMetadataByAssetId: Record<string, JsonValue>) => {
+  if (!note) throw new Error('No note to view');
+  if (!note.address) throw new Error('No address in note');
+  if (!note.value) throw new Error('No value in note');
+
+  return new NoteView({
+    address: {
+      addressView: {
+        case: 'opaque',
+        value: { address: note.address },
+      },
+    },
+    value: getValueView(note.value, denomMetadataByAssetId),
+  });
+};
+
+const getSpendView = (
+  spendPlan: SpendPlan,
+  denomMetadataByAssetId: Record<string, JsonValue>,
+): SpendView => {
+  if (!spendPlan.note?.address) throw new Error('No address in spend plan');
+
+  return new SpendView({
+    spendView: {
+      case: 'visible',
+      value: {
+        note: getNoteView(spendPlan.note, denomMetadataByAssetId),
+      },
+    },
+  });
+};
+
+const getOutputView = (
+  outputPlan: OutputPlan,
+  denomMetadataByAssetId: Record<string, JsonValue>,
+): OutputView => {
+  if (!outputPlan.destAddress) throw new Error('No destAddress in output plan');
+
+  return new OutputView({
+    outputView: {
+      case: 'visible',
+
+      value: {
+        note: {
+          value: getValueView(outputPlan.value, denomMetadataByAssetId),
+          address: {
+            addressView: {
+              case: 'opaque',
+              value: {
+                address: outputPlan.destAddress,
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+};
+
+export const getStubActionViewFromPlan =
+  (denomMetadataByAssetId: Record<string, JsonValue>) =>
+  (actionPlan: ActionPlan): ActionView => {
+    switch (actionPlan.action.case) {
+      case 'spend':
+        return new ActionView({
+          actionView: {
+            case: 'spend',
+            value: getSpendView(actionPlan.action.value, denomMetadataByAssetId),
+          },
+        });
+      case 'output':
+        return new ActionView({
+          actionView: {
+            case: 'output',
+            value: getOutputView(actionPlan.action.value, denomMetadataByAssetId),
+          },
+        });
+      case 'withdrawal':
+        /**
+         * Special case -- the `withdrawal` case in the action plan maps to the
+         * `ics20Withdrawal` case in the action view.
+         */
+        return new ActionView({
+          actionView: {
+            case: 'ics20Withdrawal',
+            value: {},
+          },
+        });
```

Comment:
@hdevalence or @turbocrime would you mind checking my logic here?

It appears from the Protobufs that in the context of an action _plan_, an ICS20 Withdrawal has a case of `withdrawal` ([code](https://github.com/penumbra-zone/penumbra/blob/main/proto/penumbra/penumbra/core/transaction/v1alpha1/transaction.proto#L255)). But in the context of an [_action_](https://github.com/penumbra-zone/penumbra/blob/main/proto/penumbra/penumbra/core/transaction/v1alpha1/transaction.proto#L98) or an [_action view_](https://github.com/penumbra-zone/penumbra/blob/main/proto/penumbra/penumbra/core/transaction/v1alpha1/transaction.proto#L188), it has a case of `ics20_withdrawal` (which is then camel-cased in JavaScript). So out of all of the different action cases, this is the only one that needs a special translation when going from action plan -> action view.

Let me know if I'm missing something, or if this is correct.

Reply:
I think this is downstream of the field naming in the .proto files and is the kind of minor inconsistency that would be good to fix asap before we stabilize formats and are stuck with it forever

Reply:
Filed an issue [here](https://github.com/penumbra-zone/penumbra/issues/3614).